PCA: Add validity test and fix several bugs

sgiraudot opened the pull request:
## Summary of Changes

When converting the deprecated API to the Eigen version a few years ago, I must have missed a division which lead to an incorrect moment matrix initialization in the case of an input with `Segment_3`, which lead to the bug https://github.com/CGAL/cgal/issues/4993.

This PR fixes it. Now the Eigen version (which is the officially documented one) works well.

This PR also adds a validity test to check that the output of each PCA variant is correct. This revealed other errors which I fixed:
- the moment matrix of `Iso_cuboid_3` was wrong too
- the overload of the variant using `Segment_2` with `CGAL::Dimension_tag<0>` was broken due to 2 inverted arguments
- the overloads of the variants using `Iso_rectangle_2` with `CGAL::Dimension_tag<1>` and `<0` were broken due to a missing argument

All is fixed.

## Release Management

* Affected package(s): PCA
* Issue(s) solved (if any): fix https://github.com/CGAL/cgal/issues/4993

afabri commented:
Can you please add test cases.   My guess is that a better testsuite had avoided the regession.

sgiraudot commented:
Alright, so I added a validity test (which was quite long to do considering the number of variants).

It actually revealed additional errors which I added to the summary of the PR. So I think it was definitely worth it, it should save us some trouble in the future.

lrineau commented:
This PR has a conflict with CGAL-5.0-branch.

maxGimeno commented:
warnings :https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-15/Principal_component_analysis/TestReport_lrineau_Fedora-Release.gz

maxGimeno commented:
warnings : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-17/Principal_component_analysis/TestReport_Blake_Windows_MSVCPreview-Release-64bits.gz

maxGimeno commented:
Runtime error : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-21/Principal_component_analysis/TestReport_lrineau_Ubuntu-Latest.gz

maxGimeno commented:
Also a warning about diagonalize_traits that is deprecated

sgiraudot commented:
The error should be fixed.

The warning seems to come from `-- Eigen3 path specified in cmake variable EIGEN3_INCLUDE_DIR is set to /usr/local/include, but that path does not contains the file signature_of_eigen3_matrix_library and is considered as invalid.` but I don't think this is related to this PR (the warning pops if `CGAL_EIGEN3_ENABLED` is not defined, which is probably due to this cmake problem).

maxGimeno commented:
Ok, thanks

lrineau commented:
> The error should be fixed.
> 
> The warning seems to come from `-- Eigen3 path specified in cmake variable EIGEN3_INCLUDE_DIR is set to /usr/local/include, but that path does not contains the file signature_of_eigen3_matrix_library and is considered as invalid.` but I don't think this is related to this PR (the warning pops if `CGAL_EIGEN3_ENABLED` is not defined, which is probably due to this cmake problem).

I think:
- our `find_package(Eigen)` should be `REQUIRED`.
- Eigen should be installed on all test platforms.

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-23.shtml
